NIFI-1273 Adding ListenRELP Processor

bbende opened the pull request:
This pull request adds a ListenRELP processor which includes refactoring code that was previously part of ListenSyslog into a reusable framework for implementing listener processors, as well as back-end code for handling the RELP protocol.

## Overivew
* The listen package in nifi-processor-utils contains the generic dispatcher/handler code along with an AbstractListenEventProcessor which is the base class for future processors that want to act as a server reading data off a channel.
* The relp package in nifi-standard-processors contains the RELP specific implementations for the "listen" classes, along with other back-end code for the RELP protocol.
* ListenRELP extends AbstractListenEventProcessor

## Testing

I tested this against rsyslog 8.x putting the following in /etc/rsyslog.conf:
    
    module(load="omrelp")
    action(type="omrelp" target="centralserv" port="2514")

With TLS:

    action(type="omrelp" target="192.168.233.153″ port="20514″ tls="on"
    tls.caCert="/home/test/cert/ca.pem"
    tls.myCert="/home/test/cert/ubuntu1-cert.pem"
    tls.myPrivKey="/home/test/cert/ubuntu1-key.pem"
    tls.authmode="name"
    tls.permittedpeer=["ubuntu-server"]
    )

RELP Reference:

    http://www.rsyslog.com/doc/relp.html
    http://www.rsyslog.com/using-tls-with-relp/

RELPFrameProducer in nifi-standard-processors src/test/java/org.apache.nifi.processors.standard.relp can also be used to act as an rsyslog server sending messages to ListenRELP.

trkurc commented:
I like the docs to help with reviewing! I good amount here to review and test!

mcgilman commented on nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/listen/dispatcher/DatagramChannelDispatcher.java:
```diff
@@ -0,0 +1,160 @@
+
+    @Override
+    public void run() {
+        final ByteBuffer buffer = bufferPool.poll();
```
Protect when buffer is null or ensure bufferPool is not empty initially?

mcgilman commented:
Spent some time reviewing the patch. Overall, I think the code looks good. Additional nit-picky item outside of my previous comment.

- I could see possibly wanting to extract the RELP utility classes into a separate artifact. Maybe we wait to shuffle this around until we have a case for it. Maybe introduce a RELP NAR if we ever get to that point.

When running ListenRELP, there were times (especially right after stopping and starting the Processor) that new messages were not picked up for some time. At some point later, a backlog of messages flooded in. It's not clear what triggered it and the backlog appeared to produce some duplicate messages.

@trkurc did you want to check this out before we proceed with merging?

trkurc commented:
@mcgilman - I have been reviewing. I should be complete tonight. Haven't found any show stoppers yet.

bbende commented:
Thanks for reviewing @mcgilman and @trkurc ! 

I pushed up a small commit that fixed a couple of minor logging issues and addressed the comment Matt made about the pool of buffers. 

Will wait to see any feedback from Tony and can address anything else as needed.

trkurc commented on nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/listen/AbstractListenEventProcessor.java:
```diff
@@ -0,0 +1,356 @@
+        if (event == null) {
+            try {
+                if (longPoll) {
+                    event = events.poll(100, TimeUnit.MILLISECONDS);
```
next chance we get, should replace this 100 with a constant.

trkurc commented on nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/listen/dispatcher/ChannelDispatcher.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * Dispatches handlers for a given channel.
+ */
+public interface ChannelDispatcher extends Runnable {
```
For extensiblity, it would be very helpful to have javadoc for this interface.

trkurc commented on nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/listen/dispatcher/DatagramChannelDispatcher.java:
```diff
@@ -0,0 +1,164 @@
+                    }
+                }
+            } catch (InterruptedException e) {
+                stopped = true;
```
when reviewing one of @olegz's patches, I thought this was a place where adding a Thread.currentThread().interrupt() made sense

trkurc commented on nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/listen/dispatcher/DatagramChannelDispatcher.java:
```diff
@@ -0,0 +1,164 @@
+            try {
+                bufferPool.put(buffer);
+            } catch (InterruptedException e) {
+                // nothing to do here
```
when reviewing one of @olegz's patches, I thought this was a another place where adding a Thread.currentThread().interrupt() made sense

trkurc commented on nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/listen/dispatcher/DatagramChannelDispatcher.java:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    @Override
+    public void completeConnection(SelectionKey key) {
```
since these two are not useful for this, should the interface ChannelDispatcher include these? Or should there be an AsyncChannelDispatcher interface that extends ChannelDispatcher?

trkurc commented on nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/listen/dispatcher/SocketChannelDispatcher.java:
```diff
@@ -0,0 +1,243 @@
+        try {
+            bufferPool.put(attachment.getByteBuffer());
+        } catch (InterruptedException e) {
+            // nothing to do here
```
like some of the above, may want to add a Thread.currentThread().interrupt();

trkurc commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ListenRELP.java:
```diff
@@ -0,0 +1,224 @@
+        "portion of one or more RELP frames. In the case where the RELP frames contain syslog messages, the " +
+        "output of this processor can be sent to a ParseSyslog processor for further processing.")
+@WritesAttributes({
+        @WritesAttribute(attribute="relp.comand", description="The command of the RELP frames."),
```
s/comand/command/

trkurc commented:
@bbende - this is an awesome contribution. most of my comments are minor and of the ilk of we should fix next chance we get, with the exception of the 'comand' typo

trixpan commented:
@bbende I have been running some tests that I usually use around reliable destinations (disconnects, repeated messages, unexpected restarts) and all worked to expected (except for the issue already noted by @mcgilman). 

Amazing code and it certainly enable NiFi to have a greater number of inbound protocols. I will port my Lumberjack spaghetti into the new and cleaner code.

bbende commented:
@trkurc thanks for the review, i liked your idea about the AsyncChannelDispatcher extending ChannelDispatcher so I went ahead and did that, and addressed the other stuff as well.

@trixpan thanks for testing this out!

As far as I can tell the scenario with the messages being resent seems to be something happening on the rsyslog side of things. I can't really come up with a pattern of what it is doing as it is not always resending the same amount of messages.

I think we are good to merge this in.

bbende commented:
Just a quick update on the scenario where some messages get resent... 

I turned on rsyslog debug and I noticed that rsyslog seemed to be queueing up responses until 128 responses were received at which point it processed all the responses and started back over, this was consistent with a long running test. So if you stop and start the processor before that, say when it has received 50 responses, it hasn't processed those yet so it thinks those 50 messages need to be resent.

There may be some advanced config in rsyslog to control this, but I am convinced we are sending the correct responses and believe we are good to merge this.

mcgilman commented:
Awesome @bbende great work! I'll go ahead and merge this in now. Thanks everyone!
